### PR TITLE
[wip] observers

### PR DESCRIPTION
**Pull request by stevekuznetsov**
todo: we need to signal to the observer process with sigint but then not cancel the test case notifier so that we can still get artifacts

### CODE REVIEW

File: pkg/steps/multi_stage.go [hunk 1]
```diff
@@ -172,14 +176,29 @@ func (s *multiStageTestStep) run(ctx context.Context) error {
...
 	if err := s.runSteps(ctx, s.pre, env, true, false); err != nil {
 		errs = append(errs, fmt.Errorf("%q pre steps failed: %w", s.name, err))
 	} else if err := s.runSteps(ctx, s.test, env, true, len(errs) != 0); err != nil {
 		errs = append(errs, fmt.Errorf("%q test steps failed: %w", s.name, err))
 	}
+	cancel() // signal to observers that we're tearing down
 	if err := s.runSteps(context.Background(), s.post, env, false, len(errs) != 0); err != nil {
 		errs = append(errs, fmt.Errorf("%q post steps failed: %w", s.name, err))
 	}
+	// we need to wait for the observers to finish so we get their jUnit
```
Comment (bbguimaraes) [inline pkg/steps/multi_stage.go:197]: I don't see it in the E2E test?

For regular steps, that is done by https://github.com/openshift/ci-tools/blob/4df23e041313f8b32bbcdb0f5967017d66ca72e9/pkg/steps/multi_stage.go#L624

File: pkg/steps/multi_stage.go [hunk 2]
```diff
@@ -172,14 +176,29 @@ func (s *multiStageTestStep) run(ctx context.Context) error {
 		return fmt.Errorf("failed to create RBAC objects: %w", err)
 	}
 	var errs []error
+	observers, err := s.generateObservers(s.observers)
+	if err != nil {
+		// if we can't even generate the Pods there's no reason to run the job
+		return err
+	}
+	observerContext, cancel := context.WithCancel(ctx)
+	observerDone := make(chan interface{})
```
Comment (bbguimaraes) [inline pkg/steps/multi_stage.go:185]: Minor point, but did you mean `struct{}`?  Even empty interfaces are a data/vtable pointer pair:



https://research.swtch.com/interfaces

**Review by bbguimaraes — COMMENTED** [review]
LGTM so far.

**Comment by stevekuznetsov** [comment]
Implements DPTP-1668

**Comment by stevekuznetsov** [comment]
Integration is somehow broken, not able to figure it out yet

File: hack/lib/integration.sh [hunk 3]
```diff
@@ -52,11 +52,11 @@ function os::integration::configresolver::start() {
 
     os::log::info "Starting the config resolver..."
     ci-operator-configresolver --config "${config}"       \
+                               --address ":8088"          \
                                --registry "${registry}"   \
                                --prow-config "${prow}"    \
                                "${flat:+--flat-registry}" \
-                               --log-level debug          \
-                               --cycle 2m >"${LOG_DIR}/configresolver.log" 2>&1 &
```
Comment (stevekuznetsov) [inline hack/lib/integration.sh:59]: `--cycle` is deprecated

File: hack/lib/integration.sh [hunk 4]
```diff
@@ -52,11 +52,11 @@ function os::integration::configresolver::start() {
 
     os::log::info "Starting the config resolver..."
     ci-operator-configresolver --config "${config}"       \
+                               --address ":8088"          \
```
Comment (stevekuznetsov) [inline hack/lib/integration.sh:55]: somehow `--address` did not get parsed when it was at the end of the flag list?

Comment (stevekuznetsov) [reply hack/lib/integration.sh:55]: we need to use somethiong *other* than `:8088` as `boskos` only knows how to serve on that port

**Comment by stevekuznetsov** [comment]
Seventeen rebases later I think this is finally current, we will see what the e2e says

**Comment by stevekuznetsov** [comment]
```
        2021/02/15 14:04:56 Acquiring leases for "with-observer"
        2021/02/15 14:04:56 Acquiring 1 lease(s) for "aws-quota-slice"
        2021/02/15 14:04:56 Acquired lease(s) for "aws-quota-slice": [3513217c-148d-46ec-be68-61bd5b379757]
        2021/02/15 14:04:56 Creating multi-stage test secret "with-observer"
        2021/02/15 14:04:56 Creating multi-stage test credentials for "with-observer"
        2021/02/15 14:05:57 Executing pod "with-observer-observer" running image "pipeline:os"
        2021/02/15 14:05:57 Executing pod "with-observer-create-kubeconfig" running image "pipeline:os"
        2021/02/15 14:06:03 Container place-entrypoint in pod with-observer-observer completed successfully
        2021/02/15 14:06:03 Container cp-secret-wrapper in pod with-observer-observer completed successfully
        2021/02/15 14:06:03 Container place-entrypoint in pod with-observer-create-kubeconfig completed successfully
        2021/02/15 14:06:03 Container cp-secret-wrapper in pod with-observer-create-kubeconfig completed successfully
        2021/02/15 14:06:28 Container sidecar in pod with-observer-create-kubeconfig completed successfully
        2021/02/15 14:06:28 Container test in pod with-observer-create-kubeconfig completed successfully
        2021/02/15 14:06:28 Pod with-observer-create-kubeconfig succeeded after 27s
        2021/02/15 14:06:38 warning: observer failed: "with-observer" pod "with-observer-observer" failed: the pod ci-op-4w019bh7/with-observer-observer failed after 41s (failed containers: ): ContainerFailed one or more containers exited
```

something no workey

**Comment by stevekuznetsov** [comment]
```
        2021/02/15 14:45:51 Acquiring leases for "with-observer"
        2021/02/15 14:45:51 Acquiring 1 lease(s) for "aws-quota-slice"
        2021/02/15 14:45:51 Acquired lease(s) for "aws-quota-slice": [da55d751-5600-46bd-97a9-df4c47990ff6]
        2021/02/15 14:45:51 Creating multi-stage test secret "with-observer"
        2021/02/15 14:45:52 Creating multi-stage test credentials for "with-observer"
        2021/02/15 14:46:53 Executing pod "with-observer-create-kubeconfig" running image "pipeline:os"
        2021/02/15 14:46:53 Executing pod "with-observer-observer" running image "pipeline:os"
        2021/02/15 14:46:58 Container place-entrypoint in pod with-observer-create-kubeconfig completed successfully
        2021/02/15 14:46:58 Container cp-secret-wrapper in pod with-observer-create-kubeconfig completed successfully
        2021/02/15 14:46:58 Container place-entrypoint in pod with-observer-observer completed successfully
        2021/02/15 14:46:58 Container cp-secret-wrapper in pod with-observer-observer completed successfully
        2021/02/15 14:47:23 Container sidecar in pod with-observer-create-kubeconfig completed successfully
        2021/02/15 14:47:23 Container test in pod with-observer-create-kubeconfig completed successfully
        2021/02/15 14:47:23 Pod with-observer-create-kubeconfig succeeded after 26s
        2021/02/15 14:47:23 Signalling observers to terminate...
        2021/02/15 14:47:23 Executing pod "with-observer-check-output" running image "pipeline:os"
        2021/02/15 14:47:29 Container place-entrypoint in pod with-observer-check-output completed successfully
        2021/02/15 14:47:29 Container cp-secret-wrapper in pod with-observer-check-output completed successfully
        2021/02/15 14:48:04 Container sidecar in pod with-observer-check-output completed successfully
        cat: /tmp/secret/cancelled: No such file or directory
        Observer was not cancelled!
        {"component":"entrypoint","error":"wrapped process failed: exit status 1","file":"prow/entrypoint/run.go:80","func":"k8s.io/test-infra/prow/entrypoint.Options.Run","level":"error","msg":"Error executing test process","severity":"error","time":"2021-02-15T22:47:58Z"}
        error: failed to execute wrapped command: exit status 1
        2021/02/15 14:48:04 Container test in pod with-observer-check-output failed, exit code 1, reason Error
```

Something about cancellation does not seem to work.

**Comment by stevekuznetsov** [comment]
@bbguimaraes I need your help on this one. I'm testing locally but can't get this to work. It doesn't seem like the `secret-wrapper` correctly puts data from the observer pod into the shared secret. Logs:

```
$ oc --context build01 --namespace ci-op-y9yd2f4v logs -c test -f with-observer-observer
+ trap cleanup EXIT
+ true
+ [[ -f /var/run/secrets/ci.openshift.io/multi-stage/kubeconfig ]]
+ echo '$KUBECONFIG does not exist, waiting...'
$KUBECONFIG does not exist, waiting...
+ sleep 1
+ true
+ [[ -f /var/run/secrets/ci.openshift.io/multi-stage/kubeconfig ]]
+ echo '$KUBECONFIG does not exist, waiting...'
$KUBECONFIG does not exist, waiting...
+ sleep 1
+ true
+ [[ -f /var/run/secrets/ci.openshift.io/multi-stage/kubeconfig ]]
+ echo '$KUBECONFIG exists'
$KUBECONFIG exists
+ break
+ echo -n waited
+ sleep 360
received signal 15, forwarding
time="2021-02-16T22:49:50Z" level=info msg="Received signal." signal=terminated
{"component":"entrypoint","file":"prow/entrypoint/run.go:169","func":"k8s.io/test-infra/prow/entrypoint.Options.ExecuteProcess","level":"error","msg":"Entrypoint received interrupt: terminated","severity":"error","time":"2021-02-16T22:49:50Z"}
^C
```

The `received signal 15, forwarding` is from the `secret-wrapper` so it's definitely running ... 

Testing like:

```
$ make install local-e2e TESTFLAGS='-run TestObservers'
```

Would appreciate your help in trying to figure out what's up/

**Comment by stevekuznetsov** [comment]
Wait, we only copy in secret state at the start of execution, right?

**Comment by stevekuznetsov** [comment]
If that's true, how does kubeconfig show up halfway through the observer run..?

**Pull request closed without merging**